### PR TITLE
Multidimensional dense workspaces

### PR DESCRIPTION
**Pull request by weiya711**
1. Add in functionality for multi-dimensional dense workspaces
2. Also add in multi-dimensional dense workspaces for the TACO command-line tool. 
3. Fix the `precompute(...)` concrete index notation transformation algorithm to be similar to the one in the [TACO workspaces paper](http://fredrikbk.com/publications/taco-workspaces.pdf).
4. Add in `ir::Stmt` `workspace_rewriter` to allow for multi-dimensional dense workspace lowering. This rewriter rewrites the lowered `ir::Stmt` to replace all workspace `GetProperty(..., Dimension)` nodes with the correct `ir::Expr` corresponding to the workspace size for that mode. 
5. Add in more workspace regression tests in `test/tests-workspaces.cpp`

This does not break any of the GPU tests (see https://github.com/tensor-compiler/taco/actions/runs/955323370)

Note: Some `*Scheduled(...)` (like `concretizeScheduled()`, `isReductionNotationScheduled()`, etc.) methods in `src/index_notation/index_notation.cpp` are not currently being used but should be useful in the future. They were originally created for another version of the `precompute(...)` transformation algorithm, but are no longer needed there.  I would still like to add them in to `master` along with this PR since there are still scenarios where the compiler needs to concretize an Einsum notation assignment with a ProvenanceGraph into its corresponding **post-scheduled** concrete index notation statement.

### CODE REVIEW

File: src/index_notation/index_notation.cpp [hunk 1]
```diff
@@ -1510,25 +1522,43 @@ IndexStmt IndexStmt::divide(IndexVar i, IndexVar i1, IndexVar i2, size_t splitFa
...
+  // TODO: need to assert they are same length
   IndexStmt transformed = *this;
   string reason;
 
-  if (i != iw) {
-    IndexVarRel rel = IndexVarRel(new PrecomputeRelNode(i, iw));
-    transformed = Transformation(AddSuchThatPredicates({rel})).apply(transformed, &reason);
-    if (!transformed.defined()) {
-      taco_uerror << reason;
+ // FIXME: need to re-enable this later
```
Comment (stephenchouca) [inline src/index_notation/index_notation.cpp:1532]: Is this fixme redundant? Not sure what needs to be re-enabled exactly.

File: src/index_notation/index_notation.cpp [hunk 2]
```diff
@@ -1510,25 +1522,43 @@ IndexStmt IndexStmt::divide(IndexVar i, IndexVar i1, IndexVar i2, size_t splitFa
   return transformed;
 }
 
-IndexStmt IndexStmt::precompute(IndexExpr expr, IndexVar i, IndexVar iw, TensorVar workspace) const {
+IndexStmt IndexStmt::precompute(IndexExpr expr, std::vector<IndexVar> i_vars,
+                                std::vector<IndexVar> iw_vars, TensorVar workspace) const {
+
+  // TODO: need to assert they are same length
```
Comment (stephenchouca) [inline src/index_notation/index_notation.cpp:1528]: Isn't this already done on line 1533?

File: src/index_notation/index_notation.cpp [hunk 3]
```diff
@@ -2339,6 +2371,67 @@ bool isReductionNotation(IndexStmt stmt, std::string* reason) {
...
+  }
+
+  if (!isValid(to<Assignment>(stmt), reason)) {
+    return false;
+  }
+
+  // Reduction notation until proved otherwise
+  bool isReduction = true;
+
+  util::ScopedMap<IndexVar,int> boundVars;  // (int) value not used
```
Comment (stephenchouca) [inline src/index_notation/index_notation.cpp:2389]: If the int value is not used, could this be replaced with a ScopedSet then?

File: src/index_notation/index_notation.cpp [hunk 4]
```diff
@@ -2597,6 +2690,241 @@ IndexStmt makeConcreteNotation(IndexStmt stmt) {
...
+              != reductionAndFreeVars.end()) {
+            stmt = forall(i, stmt);
+          }
+        }
+      }
+    }
+  }
+
+  // Replace other reductions with where and forall statements
+  struct ReplaceReductionsWithWheres : IndexNotationRewriter {
```
Comment (stephenchouca) [inline src/index_notation/index_notation.cpp:2886]: This looks unchanged from the one in `makeConcreteNotation`...maybe this should be refactored out?

File: src/lower/lowerer_impl_imperative.cpp [hunk 5]
```diff
@@ -2064,8 +2086,8 @@ vector<Stmt> LowererImplImperative::codeToInitializeTemporaryParallel(Where wher
     values = ir::Var::make(temporaryAll.getName(),
                            temporaryAll.getType().getDataType(),
                                 true, false);
-    taco_iassert(temporaryAll.getType().getOrder() == 1) << " Temporary order was "
-                                                      << temporaryAll.getType().getOrder();  // TODO
+    // taco_iassert(temporaryAll.getType().getOrder() == 1) << " Temporary order was "
```
Comment (stephenchouca) [inline src/lower/lowerer_impl_imperative.cpp:2089]: Should probably just delete this.

File: src/lower/lowerer_impl_imperative.cpp [hunk 6]
```diff
@@ -2119,8 +2141,8 @@ vector<Stmt> LowererImplImperative::codeToInitializeTemporary(Where where) {
         needComputeValues(where, temporary)) {
       values = ir::Var::make(temporary.getName(),
                              temporary.getType().getDataType(), true, false);
-      taco_iassert(temporary.getType().getOrder() == 1)
-          << " Temporary order was " << temporary.getType().getOrder();  // TODO
+      //taco_iassert(temporary.getType().getOrder() == 1)
```
Comment (stephenchouca) [inline src/lower/lowerer_impl_imperative.cpp:2144]: Should probably just delete this.

File: test/tests-scheduling-eval.cpp [hunk 7]
```diff
@@ -138,14 +153,21 @@ IndexStmt scheduleTTMCPU(IndexStmt stmt, Tensor<double> B, int CHUNK_SIZE=16, in
...
-  TensorVar w("w", Type(Float64, {Dimension(j)}), taco::dense);
-  return stmt.split(i, i1, i2, CHUNK_SIZE)
-          .reorder({i1, i2, k, l, j})
-          .precompute(precomputeExpr, j, j, w)
+  TensorVar w("w", Type(Float64, {(size_t)NUM_J}), taco::dense);
+
+  stmt = stmt.split(i, i1, i2, CHUNK_SIZE)
+    .reorder({i1, i2, k, l, j});
+
+  cout << stmt << endl;
```
Comment (stephenchouca) [inline test/tests-scheduling-eval.cpp:167]: cout's should be removed.

**Review by stephenchouca — APPROVED** [review]

File: src/index_notation/index_notation.cpp (hunk 1 above)
Comment (weiya711) [reply src/index_notation/index_notation.cpp:1532]: Yea the code below line 1532 was originally commented out and had to be added back in later. The FIXME is redundant.

File: src/index_notation/index_notation.cpp (hunk 2 above)
Comment (weiya711) [reply src/index_notation/index_notation.cpp:1528]: Yes, this comment should be removed

File: src/index_notation/index_notation.cpp (hunk 3 above)
Comment (weiya711) [reply src/index_notation/index_notation.cpp:2389]: This is the same code from `isReductionNotation(stmt)` copied. I will change it in both places.

File: src/index_notation/index_notation.cpp (hunk 4 above)
Comment (weiya711) [reply src/index_notation/index_notation.cpp:2886]: Done

File: src/lower/lowerer_impl_imperative.cpp (hunk 5 above)
Comment (weiya711) [reply src/lower/lowerer_impl_imperative.cpp:2089]: Done

File: src/lower/lowerer_impl_imperative.cpp (hunk 6 above)
Comment (weiya711) [reply src/lower/lowerer_impl_imperative.cpp:2144]: Done

File: test/tests-scheduling-eval.cpp (hunk 7 above)
Comment (weiya711) [reply test/tests-scheduling-eval.cpp:167]: Done

**Pull request merged**